src/main/java/com/faforever/api/security/FafTokenService.java

```diff
@@ -0,0 +1,102 @@
+import java.time.temporal.TemporalAmount;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+@Service
+@Slf4j
+public class FafTokenService {
+  static final String KEY_ACTION = "action";
+  static final String KEY_LIFETIME = "lifetime";
+
+  private final ObjectMapper objectMapper;
+  private final MacSigner macSigner;
+
+  public FafTokenService(ObjectMapper objectMapper, FafApiProperties properties) {
+    this.objectMapper = objectMapper;
+    this.macSigner = new MacSigner(properties.getJwt().getSecret());
+  }
+
+  /**
+   * Creates a signed token with a map of attributes and time-limited validity
+   *
+   * @see #resolveToken(FafTokenType, String)
+   */
+  @SneakyThrows
+  public String createToken(@NotNull FafTokenType type, @NotNull TemporalAmount lifetime, @NotNull Map<String, String> attributes) {
+    Assert.notNull(attributes, "Attributes map must not be null");
+    Assert.isTrue(!attributes.containsKey(KEY_ACTION), MessageFormat.format("`{0}` is a protected attributed and must not be used", KEY_ACTION));
+    Assert.isTrue(!attributes.containsKey(KEY_LIFETIME), MessageFormat.format("`{0}` is a protected attributed and must not be used", KEY_LIFETIME));
+
+    HashMap<String, String> claims = new HashMap<>(attributes);
+    claims.put(KEY_ACTION, type.toString());
+    Instant expiresAt = Instant.now().plus(lifetime);
+    claims.put(KEY_LIFETIME, expiresAt.toString());
+
+    log.debug("Creating token of type `{}` expiring at `{}` with attributes: {}", type, expiresAt, attributes);
+
+    return JwtHelper.encode(objectMapper.writeValueAsString(claims), macSigner).getEncoded();
+  }
+
+  /**
+   * Verifies a token regarding it's type, lifetime and signature
+   *
+   * @return Map of original attributes
+   * @see #createToken(FafTokenType, TemporalAmount, Map)
+   */
+  public Map<String, String> resolveToken(@NotNull FafTokenType expectedTokenType, @NotNull String token) {
+    Map<String, String> claims = null;
+
+    try {
+      claims = objectMapper.readValue(JwtHelper.decodeAndVerify(token, macSigner).getClaims(), new TypeReference<Map<String, String>>() {
+      });
+      Assert.notNull(claims, "claims must not be null");
+      Assert.isTrue(claims.containsKey(KEY_ACTION), "Token does not contain: " + KEY_ACTION);
+      Assert.isTrue(claims.containsKey(KEY_LIFETIME), "Token does not contain: " + KEY_LIFETIME);
+      FafTokenType actualTokenType = FafTokenType.valueOf(claims.get(KEY_ACTION));
+      Assert.state(Objects.equals(expectedTokenType, actualTokenType),
+        String.format("Token types do not match, expected: %s, actual: %s", expectedTokenType, actualTokenType));
+    } catch (IOException | IllegalArgumentException | IllegalStateException e) {
+      if (claims == null) {
+        log.warn("Unparseable token of expected type {}: {}", expectedTokenType, token);
+      } else {
+        log.warn("Token of expected type `{}` invalid: {}", expectedTokenType, token);
+      }
```

Comment:
Having conditions like this in a catch clause increases complexity and is destined to misbehave. We can simplify this.
And: more backticks ;) (and missing quotes)